File: test-suite-kotlin-server-generator/src/test/java/io/micronaut/openapi/test/api/ParametersControllerTest.java
package io.micronaut.openapi.test.api;

import java.math.BigDecimal;
import java.util.stream.Stream;

import io.micronaut.core.type.Argument;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.MediaType;
import io.micronaut.http.client.BlockingHttpClient;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.openapi.test.model.ColorEnum;
import io.micronaut.openapi.test.model.SendPrimitivesResponse;
import io.micronaut.runtime.server.EmbeddedServer;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.EnumSource;
import org.junit.jupiter.params.provider.MethodSource;

import jakarta.inject.Inject;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.params.provider.Arguments.arguments;

@MicronautTest
class ParametersControllerTest {

    @Inject
    EmbeddedServer server;

    @Inject
    @Client("/api")
    HttpClient reactiveClient;

    BlockingHttpClient client;

    @BeforeEach
    void setup() {
        client = reactiveClient.toBlocking();
    }

    @Test
    void testSendPrimitives() {
        HttpRequest<?> request = HttpRequest.GET("/sendPrimitives/Andrew?age=1&isPositive=true")
                .header("height", "17.3");

        var arg = Argument.of(SendPrimitivesResponse.class);
        var response = client.retrieve(request, arg, Argument.of(String.class));

        assertEquals(response.getName(), "Andrew");
        assertEquals(response.getAge(), BigDecimal.valueOf(1));
        assertEquals(response.getHeight(), Float.valueOf(17.3f));
        assertEquals(response.isPositive(), Boolean.TRUE);
    }

    @Test
    void testSendNotNullPrimitives() {
        HttpRequest<?> request = HttpRequest.GET("/sendPrimitives/Andrew");

        var arg = Argument.of(SendPrimitivesResponse.class);

        var e = assertThrows(HttpClientResponseException.class, () ->
            client.retrieve(request, arg, Argument.of(String.class)));

        assertEquals(e.getStatus(), HttpStatus.BAD_REQUEST);
        assertTrue(e.getMessage().contains("not specified"));
    }

    @Disabled
    @Test
    void testSendNullablePrimitives() {
        HttpRequest<?> request = HttpRequest.GET("/sendValidatedPrimitives")
            .accept(MediaType.TEXT_PLAIN_TYPE);
        client.retrieve(request, Argument.of(String.class), Argument.of(String.class));

        assertDoesNotThrow(() -> client.retrieve(request, Argument.of(String.class), Argument.of(String.class)));
    }

    @Test
    void testSendValidatedPrimitives() {
        HttpRequest<?> request = HttpRequest.GET("/sendValidatedPrimitives?name=Andrew&age=20&favoriteNumber=3.14&height=1.5");

        assertDoesNotThrow(() -> client.retrieve(request, Argument.of(String.class)));
    }

    static Stream<Arguments> queries() {
        return Stream.of(
            arguments("name=aa", "name: size must be between 3 and 2147483647"),
            arguments("name=123456", "name: must match \"[a-zA-Z]+\""),
            arguments("age=-2", "age: must be greater than or equal to 10"),
            arguments("age=500", "age: must be less than or equal to 200"),
            arguments("favoriteNumber=-500", "favoriteNumber: must be greater than or equal to -100.5"),
            arguments("favoriteNumber=100.6", "favoriteNumber: must be less than or equal to 100.5"),
            arguments("height=0", "height: must be greater than or equal to 0.1"),
            arguments("height=100", "height: must be less than or equal to 3")
        );
    }

    @Disabled
    @MethodSource("queries")
    @ParameterizedTest
    void testSendInvalidPrimitivesMessage(String query, String message) {

        HttpRequest<?> request = HttpRequest.GET("/sendValidatedPrimitives?" + query);

        var e = assertThrows(HttpClientResponseException.class, () ->
            client.retrieve(request, Argument.of(String.class), Argument.of(String.class)));

        assertEquals(e.getStatus(), HttpStatus.BAD_REQUEST);
        assertTrue(e.getMessage().contains(message));
    }

    @EnumSource(ColorEnum.class)
    @ParameterizedTest
    void testSendParameterEnumColor(ColorEnum color) {
        HttpRequest<?> request = HttpRequest.GET("/sendParameterEnum?colorParam=" + color.getValue());

        ColorEnum response = client.retrieve(request, Argument.of(ColorEnum.class));

        assertEquals(color, response);

        String stringResponse = client.retrieve(request, Argument.of(String.class));

        assertEquals('"' + color.getValue() + '"', stringResponse);
    }

    @Test
    void testSendDates() {
        HttpRequest<?> request =
            HttpRequest.GET("/sendDates?commitDate=2022-03-04&commitDateTime=2022-03-04T12:00:01.321Z");
        String response = client.retrieve(request, Argument.of(String.class), Argument.of(String.class));

        assertEquals("""
            {"commitDate":"2022-03-04","commitDateTime":"2022-03-04T12:00:01.321Z"}""", response);
    }

    @Test
    void testSendPageQuery() {
        HttpRequest<?> request = HttpRequest.GET("/sendPageQuery?page=2&size=20&sort=my-property,desc&sort=my-property-2");
        String response = client.retrieve(request, Argument.of(String.class), Argument.of(String.class));

        assertEquals("(page: 2, size: 20, sort: my-property(dir=DESC) my-property-2(dir=ASC))", response);
    }

    @Test
    void testSendMappedParameter() {
        var filter = "name=Andrew,age>20";

        HttpRequest<?> request = HttpRequest.GET("/sendMappedParameter")
                .header("Filter", filter);
        String response = client.retrieve(request, Argument.of(String.class), Argument.of(String.class));

        assertEquals(response, filter);
    }

    @Test
    void testSendInvalidMappedParameter() {
        HttpRequest<?> request = HttpRequest.GET("/sendMappedParameter")
            .header("Filter", "name=Andrew,age>20,something-else");
        var e = assertThrows(HttpClientResponseException.class, () ->
            client.retrieve(request, Argument.of(String.class), Argument.of(String.class)));

        assertEquals(HttpStatus.BAD_REQUEST, e.getStatus());
    }

}

File: test-suite-kotlin-kapt-server-generator/src/test/kotlin/io/micronaut/openapi/test/api/ParametersControllerTest.kt
package io.micronaut.openapi.test.api

import io.micronaut.core.type.Argument
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpStatus
import io.micronaut.http.MediaType
import io.micronaut.http.client.BlockingHttpClient
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.http.client.exceptions.HttpClientResponseException
import io.micronaut.openapi.test.model.ColorEnum
import io.micronaut.openapi.test.model.SendPrimitivesResponse
import io.micronaut.runtime.server.EmbeddedServer
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.Arguments.arguments
import org.junit.jupiter.params.provider.EnumSource
import org.junit.jupiter.params.provider.MethodSource
import java.math.BigDecimal
import java.util.stream.Stream
import kotlin.String

@MicronautTest
class ParametersControllerTest(
    var server: EmbeddedServer,
    @Client("/api")
    var reactiveClient: HttpClient,
) {

    lateinit var client: BlockingHttpClient

    @BeforeEach
    fun setup() {
        client = reactiveClient.toBlocking()
    }

    @Test
    fun testSendPrimitives() {
        val request = HttpRequest.GET<Any>("/sendPrimitives/Andrew?age=1&isPositive=true")
                .header("height", "17.3")

        val (name, age, height, isPositive) = client.retrieve(request, Argument.of(SendPrimitivesResponse::class.java), Argument.of(String::class.java))
        assertEquals(name, "Andrew")
        assertEquals(age, BigDecimal.valueOf(1))
        assertEquals(height, 17.3f)
        assertEquals(isPositive, true)
    }

    @Test
    fun testSendNotNullPrimitives() {
        val request = HttpRequest.GET<Any>("/sendPrimitives/Andrew")
        val arg = Argument.of(SendPrimitivesResponse::class.java)
        val e = assertThrows(HttpClientResponseException::class.java) { client.retrieve(request, arg, Argument.of(String::class.java)) }
        assertEquals(e.status, HttpStatus.BAD_REQUEST)
        assertTrue(e.message!!.contains("not specified"))
    }

    @Test
    fun testSendNullablePrimitives() {
        val request = HttpRequest.GET<Any>("/sendValidatedPrimitives")
                .accept(MediaType.TEXT_PLAIN_TYPE)
        client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java))
        assertDoesNotThrow { client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java)) }
    }

    @Test
    fun testSendValidatedPrimitives() {
        val request = HttpRequest.GET<Any>("/sendValidatedPrimitives?name=Andrew&age=20&favoriteNumber=3.14&height=1.5")
        assertDoesNotThrow { client.retrieve(request, Argument.of(String::class.java)) }
    }

    @MethodSource("queries")
    @ParameterizedTest
    fun testSendInvalidPrimitivesMessage(query: String, message: String) {
        val request = HttpRequest.GET<Any>("/sendValidatedPrimitives?$query")
        val e = assertThrows(HttpClientResponseException::class.java) { client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java)) }
        assertEquals(e.status, HttpStatus.BAD_REQUEST)
        assertTrue(e.message!!.contains(message))
    }

    @EnumSource(ColorEnum::class)
    @ParameterizedTest
    fun testSendParameterEnumColor(color: ColorEnum) {
        val request = HttpRequest.GET<Any>("/sendParameterEnum?colorParam=" + color.value)
        val response = client.retrieve(request, Argument.of(ColorEnum::class.java))
        assertEquals(color, response)
        val stringResponse = client.retrieve(request, Argument.of(String::class.java))
        assertEquals('"'.toString() + color.value + '"', stringResponse)
    }

    @Test
    fun testSendDates() {
        val request = HttpRequest.GET<Any>("/sendDates?commitDate=2022-03-04&commitDateTime=2022-03-04T12:00:01.321Z")
        val response = client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java))
        assertEquals("""
            {"commitDate":"2022-03-04","commitDateTime":"2022-03-04T12:00:01.321Z"}
            """.trimIndent(), response)
    }

    @Test
    fun testSendPageQuery() {
        val request = HttpRequest.GET<Any>("/sendPageQuery?page=2&size=20&sort=my-property,desc&sort=my-property-2")
        val response = client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java))
        assertEquals("(page: 2, size: 20, sort: my-property(dir=DESC) my-property-2(dir=ASC))", response)
    }

    @Test
    fun testSendMappedParameter() {
        val filter = "name=Andrew,age>20"
        val request = HttpRequest.GET<Any>("/sendMappedParameter")
                .header("Filter", filter)
        val response = client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java))
        assertEquals(response, filter)
    }

    @Test
    fun testSendInvalidMappedParameter() {
        val request = HttpRequest.GET<Any>("/sendMappedParameter")
                .header("Filter", "name=Andrew,age>20,something-else")
        val e = assertThrows(HttpClientResponseException::class.java) { client.retrieve(request, Argument.of(String::class.java), Argument.of(String::class.java)) }
        assertEquals(HttpStatus.BAD_REQUEST, e.status)
    }

    companion object {
        @JvmStatic
        fun queries(): Stream<Arguments> {
            return Stream.of(
                    arguments("name=aa", "name: size must be between 3 and 2147483647"),
                    arguments("name=123456", "name: must match \\\"[a-zA-Z]+\\\""),
                    arguments("age=-2", "age: must be greater than or equal to 10"),
                    arguments("age=500", "age: must be less than or equal to 200"),
                    arguments("favoriteNumber=-500", "favoriteNumber: must be greater than or equal to -100.5"),
                    arguments("favoriteNumber=100.6", "favoriteNumber: must be less than or equal to 100.5"),
                    arguments("height=0", "height: must be greater than or equal to 0.1"),
                    arguments("height=100", "height: must be less than or equal to 3")
            )
        }
    }
}
